monkey/infection_monkey/utils/signal_handler.py

```diff
@@ -1,50 +1,77 @@
 import logging
 import signal
+from typing import Optional
 
 from infection_monkey.i_master import IMaster
 from infection_monkey.utils.environment import is_windows_os
 
 logger = logging.getLogger(__name__)
 
 
+_signal_handler = None
+
+
 class StopSignalHandler:
     def __init__(self, master: IMaster):
         self._master = master
 
-    def handle_posix_signals(self, signum: int, _):
-        self._handle_signal(signum, False)
+    # Windows won't let us correctly deregister a method, but Callables and closures work.
+    def __call__(self, signum: int, *args) -> Optional[bool]:
+        if is_windows_os():
+            return self.handle_windows_signals(signum)
+        else:
+            self.handle_posix_signals(signum, args)
 
-    def handle_windows_signals(self, signum: int):
+    def handle_windows_signals(self, signum: int) -> bool:
         import win32con
 
         if signum in {win32con.CTRL_C_EVENT, win32con.CTRL_BREAK_EVENT}:
-            self._handle_signal(signum, False)
+            self._terminate_master(signum, False)
             return True
 
         if signum == win32con.CTRL_CLOSE_EVENT:
             # After the signal handler returns True, the OS will forcefully kill the process.
             # Calling self._handle_signal() with block=True to give the master a chance to
             # gracefully shut down. Note that the OS has a timeout that will forcefully kill the
             # process if this handler hasn't returned in time.
-            self._handle_signal(signum, True)
+            self._terminate_master(signum, True)
             return True
 
         return False
 
-    def _handle_signal(self, signum: int, block: bool):
+    def handle_posix_signals(self, signum: int, *_):
+        self._terminate_master(signum, False)
+
+    def _terminate_master(self, signum: int, block: bool):
         logger.info(f"The Monkey Agent received signal {signum}")
         self._master.terminate(block)
 
 
 def register_signal_handlers(master: IMaster):
-    stop_signal_handler = StopSignalHandler(master)
+    global _signal_handler
+    _signal_handler = StopSignalHandler(master)
 
     if is_windows_os():
         import win32api
 
         # CTRL_CLOSE_EVENT signal has a timeout of 5000ms,
         # after that OS will forcefully kill the process
-        win32api.SetConsoleCtrlHandler(stop_signal_handler.handle_windows_signals, True)
+        win32api.SetConsoleCtrlHandler(_signal_handler, True)
+    else:
+        signal.signal(signal.SIGINT, _signal_handler)
+        signal.signal(signal.SIGTERM, _signal_handler)
```

Comment:
For linux, `_signal_handler` will be `None` (because of line 23), i.e. we're setting the handler of these two signals to `None`. How does this work?

Reply:
`_signal_handler` is a function reference that's being passed to `signal.signal()`. We're not passing the return value resulting from a a function call, we're passing the function itself.

Reply:
Continued [here](https://github.com/guardicore/monkey/pull/1692/files#r798274611)